File: frontend/src/api/opportunity.ts
import request from './request'

export interface OppListItem {
  id: string
  opp_name: string
  company_name: string
  stage: string
  certification_type: string
  estimated_amount: number | null
  expected_close_date: string | null
  win_probability: number | null
  assigned_to: string
  lead_id: string | null
  customer_id: string | null
  created_at: string
}

export interface OppDetail extends OppListItem {
  competitor: string | null
  loss_reason: string | null
  updated_at: string
  created_by: string | null
}

export interface OppFollowUp {
  id: string
  opp_id: string
  follow_type: string
  content: string
  next_date: string | null
  created_by: string | null
  created_at: string
}

export const oppApi = {
  list(params: { page?: number; page_size?: number; stage?: string; keyword?: string }) {
    return request.get<any, { code: number; data: OppListItem[]; total: number }>('/opportunities', { params })
  },
  get(id: string) {
    return request.get<any, { code: number; data: OppDetail }>(`/opportunities/${id}`)
  },
  create(data: Record<string, unknown>) {
    return request.post<any, { code: number; data: OppDetail }>('/opportunities', data)
  },
  update(id: string, data: Record<string, unknown>) {
    return request.put<any, { code: number; data: OppDetail }>(`/opportunities/${id}`, data)
  },
  delete(id: string) {
    return request.delete(`/opportunities/${id}`)
  },
  listFollowUps(oppId: string) {
    return request.get<any, { code: number; data: OppFollowUp[] }>(`/opportunities/${oppId}/follow-ups`)
  },
  addFollowUp(oppId: string, data: Record<string, unknown>) {
    return request.post(`/opportunities/${oppId}/follow-ups`, data)
  },
}

File: backend/app/schemas/opportunity.py
import uuid
from datetime import datetime, date
from decimal import Decimal
from pydantic import BaseModel, ConfigDict


# ── 商机跟进记录 ─────────────────────────────────────────────
class OppFollowUpCreate(BaseModel):
    follow_type: str
    content: str
    next_date: str | None = None


class OppFollowUpOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    opp_id: uuid.UUID
    follow_type: str
    content: str
    next_date: str | None
    created_by: uuid.UUID | None
    created_at: datetime


# ── 商机 ─────────────────────────────────────────────────────
class OppBase(BaseModel):
    opp_name: str
    lead_id: uuid.UUID | None = None
    customer_id: uuid.UUID | None = None
    company_name: str
    stage: str = "初步接触"
    certification_type: str
    estimated_amount: Decimal | None = None
    expected_close_date: date | None = None
    win_probability: int | None = None
    competitor: str | None = None
    loss_reason: str | None = None
    assigned_to: uuid.UUID


class OppCreate(OppBase):
    pass


class OppUpdate(BaseModel):
    opp_name: str | None = None
    lead_id: uuid.UUID | None = None
    customer_id: uuid.UUID | None = None
    company_name: str | None = None
    stage: str | None = None
    certification_type: str | None = None
    estimated_amount: Decimal | None = None
    expected_close_date: date | None = None
    win_probability: int | None = None
    competitor: str | None = None
    loss_reason: str | None = None
    assigned_to: uuid.UUID | None = None


class OppOut(OppBase):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    created_by: uuid.UUID | None
    created_at: datetime
    updated_at: datetime


class OppListItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: uuid.UUID
    opp_name: str
    company_name: str
    stage: str
    certification_type: str
    estimated_amount: Decimal | None
    expected_close_date: date | None
    win_probability: int | None
    assigned_to: uuid.UUID
    lead_id: uuid.UUID | None
    customer_id: uuid.UUID | None
    created_at: datetime
